[KOGITO-7253] Workflow error handling guide (#31)
kiegroup/kogito-docs: serverlessworkflow/modules/ROOT/pages/core/understanding-workflow-error-handling.adoc

= Understanding workflow error handling

> This guide explains how to handle errors that might occur while executing workflow states. Although a non desirable situation, Workflow states, as part of their interaction with external systems, might fail. For these situations, Serverless Workflow specification provides an link:{spec_doc_url}#workflow-error-handling[error handling] mechanism which allows users to deal with these rainy day circumstances.

Reviewer: I didn't understand the first part correctly.
Handle errors is how to handle errors or handle errors is a method from serverless workflow? (a question that could be asked)

An error is an exceptional occurrence that alters the regular Workflow execution sequence. When an error occurs, rather than transitioning to the state that would have been executed in case of a successful execution, the flow, if configured to do so, jumps instead to an alternative state, potentially capable of dealing with it. Note that error definition is always optional, if no error handling is defined, the Workflow execution will be aborted when something goes wrong. If you are a software developer, you can can think of Workflow error handling  as a `try-catch` or a `goto` construct.

== Error definition

> An link:{spec_doc_url}#error-definition[Error definition] in Serverless Workflow is composed of a name and an code. The name is a natural language description of the error, for example: “wrong parameter”, while the code is a vendor specific parameter that helps the implementor to identify the error. In the case of Kogito, the code is mandatory and consist of the fully qualified Java class name of an link:https://docs.oracle.com/javase/tutorial/essential/exceptions/index.html[Exception type] that might be thrown by the state.

Reviewer: ```suggestion
An link:{spec_doc_url}#error-definition[Error definition] in Serverless Workflow is composed of a name and an code. The name is a natural language description of the error, for example: 'wrong parameter', while the code is a vendor specific parameter that helps the implementor to identify the error. In the case of Kogito, the code is mandatory and consist of the fully qualified Java class name of an link:https://docs.oracle.com/javase/tutorial/essential/exceptions/index.html[Exception type] that might be thrown by the state.
```

> The algorithm used by Kogito to determine if an exception should be handled by an error definition is:

Reviewer: shouldn't be steps instead algorithm?

Author: Well, al algorithm is another name for a set of rules, isnt it? ;)

Author: ok, changed to steps

> * it checks if the type of the Java Exception being thrown is an instance (which includes any descendant) of the Java Exception specified in `code` property of the error definition. 

Reviewer: ```suggestion
* It checks if the type of the Java Exception being thrown is an instance (which includes any descendant) of the Java Exception specified in `code` property of the error definition. 
```

> * if this not the case, it recursively checks if the root cause of the Java Exception is an instance (which includes any descendant) of the Java Exception specified in `code` property of the error definition.

Reviewer: checks if the root cause.... is duplicated from the first bullet, maybe rephrase it?

Author: yes, I normally dislike duplication, but in this case,  rephrasing the second bullet might indicate the procedure is different for the root cause, which is not. 
What about "if this not the case, it recursively performs the previous checks over the the chains of root causes of the Java Exception"?

Author: changed, please take a look @spolti

== Error handling example 

> Lets illustrate error handling with this link:{kogito_sw_examples_url}/serverless-workflow-error-quarkus[Serverless Workflow Application]. 

Author: Too many lets, isnt it? I think the fact the diagram will appear below this paragraph make this lets take a look kind of implicit, wdyt?

> Lets illustrate error handling with this link:{kogito_sw_examples_url}/serverless-workflow-error-quarkus[Serverless Workflow Application]. 

Reviewer: maybe, yes.

This functionality is achieved by:

* Defining a function that invokes the `isEven` method of `org.kie.kogito.examples.EvenService` class. This function use link:{spec_doc_url}#defining-custom-function-types[custom type] to enable the workflow to invoke java methods.
+
[source, json]
----
{
    "name": "isEven",
    "type": "custom",
    "operation": "service:java:org.kie.kogito.examples.EvenService::isEven"
}
----
* Adding an error definition that handles any runtime exception at the beginning of the link:{kogito_sw_examples_url}/serverless-workflow-error-quarkus/src/main/resources/error.sw.json[workflow]. 
+
[source, json]
----
"errors": [
    {
      "name": "odd number",
      "code": "java.lang.RuntimeException"
     }
  ]
----

> * Referring this error definition in the 'checkEven' state to tell the workflow that if this error occurs it should transition to `odd` state rather than the regular transition to `even` state. Since `IllegalArgumentException` is a child of `RuntimeException`, it will be handled by the `errorRef` if thrown by `isEven` function.

Reviewer 2: ```suggestion
* Referring to this error definition in the 'checkEven' state to tell the workflow that if this error occurs, it should transition to an `odd` state rather than the regular transition to the `even` state. Since `IllegalArgumentException` is a child of `RuntimeException`, it will be handled by the `errorRef` if thrown by the `isEven` function.
```

// links
:java_exception_url: https://docs.oracle.com/javase/tutorial/essential/exceptions/index.html

> This document describes how you can handle the errors that might occur when you execute your Serverless Workflow states. 

Reviewer 2: Not sure if "when you execute your Serverless Workflow states" passes the right message. It's not the user who executes the states; it's the engine. Too bad if left as "when the Serverless Workflow states are executed"? I know that's passive voice, but it makes more sense to me.

[[con-error-definition]]
== Error definition

An link:{spec_doc_url}#error-definition[error definition] in Serverless Workflow is composed of a `name` and `code` parameters. The `name` is a short description of an error, such as `wrong parameter`, and the `code` parameter helps the implementation to identify the error.

[NOTE]
====
In Kogito, the `code` parameter is mandatory, which consists of the fully qualified Java class name of an link:{java_exception_url}[exception type].
====

To determine if an exception can be handled by an error definition, the Serverless Workflow first checks whether or not the type of the Java exception is an instance of the exception that is specified in the error definition. Otherwise, the Serverless Workflow recursively applies the same check to the root of the original Java exception till the exception chain is exhausted.

[[ref-example-error-handling]]
== Example of error handling

The workflow in the link:{kogito_sw_examples_url}/serverless-workflow-error-quarkus[`serverless-workflow-error-quarkus`] example application describes the error handling feature as follows: 

.Example of error handling
image::core/error_handling.png[]

> In the `serverless-workflow-error-quarkus` example application, the workflow accepts an integer number as an input model. When the workflow starts, it invokes the `isEven` Java service, which passes the the defined integer number as a parameter. The link:{kogito_sw_examples_url}serverless-workflow-error-quarkus/src/main/java/org/kie/kogito/examples/EvenService.java[`isEven` Java service] throws an `IllegalArgumentException` exception if the passed integer number is odd. This means that if the integer number is even, the Serverless Workflow transitions to the even state, however, if the integer number is odd, the Serverless Workflow transitions to the odd state, resulting in a different model output.

Reviewer 2: ```suggestion
In the `serverless-workflow-error-quarkus` example application, the workflow accepts an integer number as an input model. When the workflow starts, it invokes the `isEven` Java service, which passes the defined integer number as a parameter. The link:{kogito_sw_examples_url}serverless-workflow-error-quarkus/src/main/java/org/kie/kogito/examples/EvenService.java[`isEven` Java service] throws an `IllegalArgumentException` exception if the passed integer number is odd. This means that if the integer number is even, the Serverless Workflow transitions to the `even` state, however, if the integer number is odd, the Serverless Workflow transitions to the `odd` state, resulting in a different model output.
```

Small change just to highlight that even and odd, in this context, are state names.

Author: I already changed
